lib/Dialect/FIRRTL/Transforms/ResolveTraces.cpp

```diff
@@ -160,60 +241,55 @@ struct ResolveTracesPass : public ResolveTracesBase<ResolveTracesPass> {
       type = is.getTargetResult().getType();
     else {
       if (op->getNumResults() != 1)
-        return false;
+        return std::nullopt;
       type = op->getResultTypes().front();
     }
 
     auto baseType = getBaseType(type_cast<FIRRTLType>(type));
-    return updateTargetImpl(anno, module, baseType, name);
+    return updateTargetImpl(anno, module, baseType,
+                            hw::InnerRefAttr::get(module.getModuleNameAttr(),
+                                                  innerSym.getSymName()),
+                            OpAnnoTarget(op));
   }
 
   /// Add a "target" field to an Annotation on a Module that indicates the
   /// current location of the module.  This will be local or non-local depending
   /// on the Annotation.
-  bool updateModuleTarget(FModuleLike &module, Annotation &anno) {
-
-    if (!anno.isClass(traceAnnoClass))
-      return false;
-
-    LLVM_DEBUG(llvm::dbgs() << "  - before: " << anno.getDict() << "\n");
-
-    SmallString<64> newTarget("~");
-    newTarget.append(module->getParentOfType<CircuitOp>().getName());
-    newTarget.append("|");
+  std::optional<AnnoPathValue> updateModuleTarget(FModuleLike &module,
+                                                  Annotation &anno) {
+    SmallVector<InstanceOp> instances;
 
     if (auto nla = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal")) {
       hw::HierPathOp path = nlaTable->getNLA(nla.getAttr());
       for (auto part : path.getNamepath().getValue().drop_back()) {
         auto inst = cast<hw::InnerRefAttr>(part);
-        newTarget.append(inst.getModule());
-        newTarget.append("/");
-        newTarget.append(inst.getName());
-        newTarget.append(":");
+        instances.push_back(dyn_cast<InstanceOp>(
+            istc->getInnerSymbolTable(nlaTable->getModule(inst.getModule()))
+                .lookupOp(inst.getName())));
       }
     }
-    newTarget.append(module.getModuleName());
 
-    anno.setMember("target", StringAttr::get(module->getContext(), newTarget));
+    AnnoPathValue path(instances, OpAnnoTarget(module), 0);
 
-    LLVM_DEBUG(llvm::dbgs() << "    after:  " << anno.getDict() << "\n");
-
-    return true;
+    return path;
   }
 };
 
 void ResolveTracesPass::runOnOperation() {
   LLVM_DEBUG(
       llvm::dbgs() << "==----- Running ResolveTraces "
-                      "-----------------------------------------------===\n");
-
-  // Populate the NLA Table.
-  nlaTable = &getAnalysis<NLATable>();
+                      "-----------------------------------------------===\n"
+                   << "Annotation Modifications:\n");
 
   // Grab the circuit (as this is used a few times below).
   CircuitOp circuit = getOperation();
   MLIRContext *context = circuit.getContext();
 
+  // Populate pointer datastructures.
+  nlaTable = &getAnalysis<NLATable>();
+  hw::InnerSymbolTableCollection innerSymbolTableCollection(circuit);
+  istc = &innerSymbolTableCollection;
```

Comment:
Can we use `istc = &getAnalysis<hw::InnerSymbolTableCollection>(circuit);`?

Reply:
Just `istc = &getAnalysis<hw::InnerSymbolTableCollection>()` works. 👍  Good suggestion!